### ISSUE TITLE
(edgeHub) Connect / GET Twin Possible Blocking Behavior; Disconnected Gateway Scenario

### ISSUE THREAD

**Issue by jackt-moran** [Azure/iotedge#4815]
## Expected Behavior
When an IoT Edge Gateway device loses internet connection leaf devices on the local network should be able to connect, GET their twin (the version local to edgeHub), and begin sending messages to edgeHub. While the Gateway device has no internet connection it should still be able to process all incoming messages from leaf devices, in real time, so that all real time applications on IoT Edge continue to work in offline scenarios.

## Current Behavior
When a leaf device on the Gateway's local network tries to connect to, GET twin from, then start sending messages to, a disconnected IoT Edge Gateway, edgeHub appears to get hung up trying to establish a connection to IoT Hub and/or trying to GET the latest twin in IoT Hub.

## Steps to Reproduce
Provide a detailed set of steps to reproduce the bug.
1. Set up and IoT Edge device as a gateway with a single leaf device.
2. Deploy some sample template to the gateway. For instance, a single module that logs incoming messages.
3. Ensure the IoT Edge device is running and the leaf device is not yet running.
4. Simulate extended offline operation for the IoT Edge device (by using firewall rules, or turning off wifi, for example).
5. Run the leaf device where it tries to connect and then start sending messages to edgeHub.

### Output of `iotedge check`
The errors here are expected given that the edge machine is offline.
<details>
<summary>Click here</summary>

```


Configuration checks
--------------------
√ config.yaml is well-formed - OK
√ config.yaml has well-formed connection string - OK
√ container engine is installed and functional - OK
√ config.yaml has correct hostname - OK
× config.yaml has correct URIs for daemon mgmt endpoint - Error
    Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
    docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
    See 'docker run --help'.
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
‼ latest security daemon - Warning
    Error while fetching latest versions of edge components: could not send HTTP request
        caused by: error trying to connect: Connection timed out (os error 110)
√ host time is close to real time - OK
× container time is close to host time - Error
    Could not query local time inside container
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
√ DNS server - OK
√ production readiness: certificates - OK
√ production readiness: container engine - OK
√ production readiness: logs policy - OK
√ production readiness: Edge Agent's storage directory is persisted on the host filesystem - OK
√ production readiness: Edge Hub's storage directory is persisted on the host filesystem - OK

Connectivity checks
-------------------
× host can connect to and perform TLS handshake with IoT Hub AMQP port - Error
    Could not connect to <omitted>.azure-devices.net:5671
        caused by: connection timed out
× host can connect to and perform TLS handshake with IoT Hub HTTPS / WebSockets port - Error
    Could not connect to <omitted>.azure-devices.net:443
        caused by: connection timed out
× host can connect to and perform TLS handshake with IoT Hub MQTT port - Error
    Could not connect to <omitted>.azure-devices.net:8883
        caused by: connection timed out
× container on the default network can connect to IoT Hub AMQP port - Error
    Container on the default network could not connect to <omitted>.azure-devices.net:5671
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
× container on the default network can connect to IoT Hub HTTPS / WebSockets port - Error
    Container on the default network could not connect to <omitted>.azure-devices.net:443
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
× container on the default network can connect to IoT Hub MQTT port - Error
    Container on the default network could not connect to <omitted>.azure-devices.net:8883
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
× container on the IoT Edge module network can connect to IoT Hub AMQP port - Error
    Container on the azure-iot-edge network could not connect to <omitted>.azure-devices.net:5671
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
× container on the IoT Edge module network can connect to IoT Hub HTTPS / WebSockets port - Error
    Container on the azure-iot-edge network could not connect to <omitted>.azure-devices.net:443
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.
× container on the IoT Edge module network can connect to IoT Hub MQTT port - Error
    Container on the azure-iot-edge network could not connect to <omitted>.azure-devices.net:8883
        caused by: docker returned exit code: 125, stderr = Unable to find image 'mcr.microsoft.com/azureiotedge-diagnostics:1.1.1' locally
                   docker: Error response from daemon: Get https://mcr.microsoft.com/v2/: net/http: request canceled while waiting for connection (Client.Timeout exceeded while awaiting headers).
                   See 'docker run --help'.

11 check(s) succeeded.
1 check(s) raised warnings.
11 check(s) raised errors.

```
</details>

### Device Information
* Host OS [e.g. Ubuntu 18.04, Windows Server IoT 2019]: Ubuntu 18.04
* Architecture [e.g. amd64, arm32, arm64]: amd64
* Container OS [e.g. Linux containers, Windows containers]: Linux containers

### Runtime Versions
* aziot-edged [run `iotedge version`]: 1.1.1
* Edge Agent [image tag (e.g. 1.0.0)]: 1.1.1
* Edge Hub [image tag (e.g. 1.0.0)]: 1.1.1
* Docker/Moby [run `docker version`]: 
* 
Client:
 Version:           20.10.5+azure
 API version:       1.41
 Go version:        go1.13.15
 Git commit:        55c4c88966a912ddb365e2d73a4969e700fc458f
 Built:             Tue Mar  2 17:58:58 2021
 OS/Arch:           linux/amd64
 Context:           default
 Experimental:      true

Server:
 Engine:
  Version:          20.10.5+azure
  API version:      1.41 (minimum version 1.12)
  Go version:       go1.13.15
  Git commit:       363e9a88a11be517d9e8c65c998ff56f774eb4dc
  Built:            Thu Feb 25 06:27:18 2021
  OS/Arch:          linux/amd64
  Experimental:     false
 containerd:
  Version:          1.4.4+azure
  GitCommit:        05f951a3781f4f2c1911b05e61c160e9c30eaa8e
 runc:
  Version:          1.0.0-rc93
  GitCommit:        12644e614e25b05da6fd08a38ffa0cfe1903fdec
 docker-init:
  Version:          0.19.0
  GitCommit:

## Logs

I'll keep the logs brief until I hear back with some confirmation (or not) of the path I'm going down. Please let me know if more detail is needed to confirm the possibility of the issue I'm experiencing.

<details>
<summary>aziot-edged logs</summary>

```

<Paste here between the triple backticks>

```
</details>

<details>
<summary>edge-agent logs</summary>

```

<Paste here between the triple backticks>

```
</details>

As far as I can tell, the key here are the edgeHub logs. Of note are the timestamps, in particular: 2021-04-08 00:34:10.131. At this time, edgeHub tries to reconnect, and all incoming messages pause while this operation is running. Once the operation fails, after an apparent 60 seconds of blocking, a few dozen messages come through until it begins trying to connect again... and so it continues, seemingly indefinitely. For reference, this system usually has 10K events per minute (compared to the dozens per minute shown below). So there is clearly a degradation of service.
<details>
<summary>edge-hub logs</summary>

```
<7> 2021-04-08 00:33:02.078 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.IDeviceScopeIdentitiesCache] - Getting service identity for MTC01_DATAREADER
<7> 2021-04-08 00:33:02.078 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.DeviceScopeAuthenticator] - Client MTC01_DATAREADER in device scope reauthenticated locally.
<7> 2021-04-08 00:33:02.078 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Authenticator] - Client MTC01_DATAREADER re-authentication successful
<7> 2021-04-08 00:33:02.078 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionReauthenticator] - Reauthenticated client MTC01_DATAREADER successfully
<7> 2021-04-08 00:33:08.679 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.IDeviceScopeIdentitiesCache] - Getting service identity for MTC01_DATAREADER
<6> 2021-04-08 00:33:08.679 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.DeviceScopeAuthenticator] - Client MTC01_DATAREADER in device scope authenticated locally.
<7> 2021-04-08 00:33:08.680 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Authenticator] - Client MTC01_DATAREADER authentication successful
<6> 2021-04-08 00:33:08.680 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceIdentityProvider] - Successfully generated identity for clientId MTC01_DATAREADER and username <omitted>.azure-devices.net/MTC01_DATAREADER/?api-version=2019-10-01&DeviceClientType=azure-iot-device-iothub-py%2F2.5.1%283.7.5%3BLinux%20%23144-Ubuntu%20SMP%20Fri%20Mar%2019%2014%3A12%3A35%20UTC%202021%3Bx86_64%29
<6> 2021-04-08 00:33:08.680 +00:00 [INF] [EdgeHub] - ClientAuthenticated, MTC01_DATAREADER, 38ba31fa
<6> 2021-04-08 00:33:08.680 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - New device connection for device MTC01_DATAREADER
<4> 2021-04-08 00:33:08.680 +00:00 [WRN] [EdgeHub] - Closing connection for device: MTC01_DATAREADER, Microsoft.Azure.Devices.Edge.Hub.Core.MultipleConnectionsException: Multiple connections detected for device MTC01_DATAREADER,
<6> 2021-04-08 00:33:08.680 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Disposing MessagingServiceClient for device Id MTC01_DATAREADER because of exception - Microsoft.Azure.Devices.Edge.Hub.Core.MultipleConnectionsException: Multiple connections detected for device MTC01_DATAREADER
<6> 2021-04-08 00:33:08.680 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceProxy] - Closing device proxy for device Id MTC01_DATAREADER
<6> 2021-04-08 00:33:08.680 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Client MTC01_DATAREADER connected to edgeHub, processing existing subscriptions.
<7> 2021-04-08 00:33:08.680 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.EdgeHubConnection] - Updating device MTC01_DATAREADER connection status to Connected
<6> 2021-04-08 00:33:08.681 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Bind device proxy for device MTC01_DATAREADER
<6> 2021-04-08 00:33:08.681 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Binding message channel for device Id MTC01_DATAREADER
<7> 2021-04-08 00:33:08.685 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.SessionStatePersistenceProvider] - Processing session subscriptions DesiredPropertyUpdates, TwinResponse for client MTC01_DATAREADER:
<6> 2021-04-08 00:33:08.686 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing subscriptions DesiredPropertyUpdates, TwinResponse for client MTC01_DATAREADER.
<6> 2021-04-08 00:33:08.686 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing pending subscriptions for MTC01_DATAREADER
<6> 2021-04-08 00:33:08.686 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.SessionStateStoragePersistenceProvider] - Set subscriptions from session state for MTC01_DATAREADER
<7> 2021-04-08 00:33:08.686 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - GetTwin call received from MTC01_DATAREADER
<7> 2021-04-08 00:33:08.686 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER
<7> 2021-04-08 00:33:10.058 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected and reason Retry_Expired for MTC01_DATAREADER
<7> 2021-04-08 00:33:10.061 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Operation OpenAsync timed out for MTC01_DATAREADER
<4> 2021-04-08 00:33:10.061 +00:00 [WRN] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Error creating cloud connection for client MTC01_DATAREADER
<6> 2021-04-08 00:33:10.062 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Error getting cloud connection for device MTC01_DATAREADER
<7> 2021-04-08 00:33:10.063 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER from store
<6> 2021-04-08 00:33:10.062 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Error getting cloud connection for device MTC01_DATAREADER
<6> 2021-04-08 00:33:10.063 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing pending subscriptions for MTC01_DATAREADER, but no cloud proxy was found
<7> 2021-04-08 00:33:10.063 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing subscription DesiredPropertyUpdates for client MTC01_DATAREADER.
<7> 2021-04-08 00:33:10.063 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing subscription TwinResponse for client MTC01_DATAREADER.
<6> 2021-04-08 00:33:10.062 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Error getting cloud connection for device MTC01_DATAREADER
<7> 2021-04-08 00:33:10.063 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing subscription DesiredPropertyUpdates for client MTC01_DATAREADER.
<7> 2021-04-08 00:33:10.063 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.SubscriptionProcessor] - Processing subscription TwinResponse for client MTC01_DATAREADER.
<6> 2021-04-08 00:33:10.063 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Error getting cloud connection for device MTC01_DATAREADER
<7> 2021-04-08 00:33:10.064 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER from store
<7> 2021-04-08 00:33:10.081 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceProxy] - Sent twin update to MTC01_DATAREADER
<7> 2021-04-08 00:33:10.081 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Processed GetTwin for MTC01_DATAREADER
<7> 2021-04-08 00:33:10.081 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Getting twin for device Id MTC01_DATAREADER
<7> 2021-04-08 00:33:10.083 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceProxy] - Sent twin update to MTC01_DATAREADER
<7> 2021-04-08 00:33:10.083 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Processed GetTwin for MTC01_DATAREADER
<7> 2021-04-08 00:33:10.083 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Getting twin for device Id MTC01_DATAREADER
<7> 2021-04-08 00:33:10.083 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - GetTwin call received from MTC01_DATAREADER
<7> 2021-04-08 00:33:10.083 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER
<7> 2021-04-08 00:33:10.083 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Creating cloud connection for client MTC01_DATAREADER using EdgeHub credentials
<6> 2021-04-08 00:33:10.083 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Attempting to connect to IoT Hub for client MTC01_DATAREADER via AMQP...
<7> 2021-04-08 00:34:10.086 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected and reason Retry_Expired for MTC01_DATAREADER
<7> 2021-04-08 00:34:10.087 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Operation OpenAsync timed out for MTC01_DATAREADER
<4> 2021-04-08 00:34:10.087 +00:00 [WRN] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Error creating cloud connection for client MTC01_DATAREADER
<6> 2021-04-08 00:34:10.089 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Error getting cloud connection for device MTC01_DATAREADER
<7> 2021-04-08 00:34:10.090 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER from store
<7> 2021-04-08 00:34:10.108 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceProxy] - Sent twin update to MTC01_DATAREADER
<7> 2021-04-08 00:34:10.108 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Processed GetTwin for MTC01_DATAREADER
<7> 2021-04-08 00:34:10.108 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Getting twin for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.108 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.108 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.108 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.109 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.109 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.109 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.110 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.110 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.110 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.111 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.111 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.111 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.113 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.113 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.113 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.114 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.114 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.114 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.115 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.115 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.115 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.116 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.116 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.116 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.117 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.117 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.117 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.119 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.119 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.119 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.120 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.120 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.120 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.121 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.121 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.121 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.122 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.122 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.122 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.123 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.123 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.123 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.124 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:34:10.124 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:34:10.125 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:34:10.131 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - GetTwin call received from MTC01_DATAREADER
<7> 2021-04-08 00:34:10.131 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER
<7> 2021-04-08 00:34:10.131 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Creating cloud connection for client MTC01_DATAREADER using EdgeHub credentials
<6> 2021-04-08 00:34:10.131 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Attempting to connect to IoT Hub for client MTC01_DATAREADER via AMQP...
<7> 2021-04-08 00:35:10.130 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Received connection status changed callback with connection status Disconnected and reason Retry_Expired for MTC01_DATAREADER
<7> 2021-04-08 00:35:10.131 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.ConnectivityAwareClient] - Operation OpenAsync timed out for MTC01_DATAREADER
<4> 2021-04-08 00:35:10.131 +00:00 [WRN] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Error creating cloud connection for client MTC01_DATAREADER
<6> 2021-04-08 00:35:10.132 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.Core.ConnectionManager] - Error getting cloud connection for device MTC01_DATAREADER
<7> 2021-04-08 00:35:10.133 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER from store
<7> 2021-04-08 00:35:10.151 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.DeviceProxy] - Sent twin update to MTC01_DATAREADER
<7> 2021-04-08 00:35:10.151 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Device.DeviceMessageHandler] - Processed GetTwin for MTC01_DATAREADER
<7> 2021-04-08 00:35:10.151 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Getting twin for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.151 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.151 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.151 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.152 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.152 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.152 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.155 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.155 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.155 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.156 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.156 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.156 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.156 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.156 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.156 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.157 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.157 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.157 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.158 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.158 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.159 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.159 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.159 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.159 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.160 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.160 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.160 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.161 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.161 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.161 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.162 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.162 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.162 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.164 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.164 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.164 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.166 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.166 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.166 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.167 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.167 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.167 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.167 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessageAddressConverter] - Message address devices/MTC01_DATAREADER/messages/events/ matches more than one topic. Picking first matching topic.
<7> 2021-04-08 00:35:10.167 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Mqtt.MessagingServiceClient] - Processing message for device Id MTC01_DATAREADER
<7> 2021-04-08 00:35:10.167 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - Received message from MTC01 DATAREADER
<7> 2021-04-08 00:35:10.168 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Routing.RoutingEdgeHub] - GetTwin call received from MTC01_DATAREADER
<7> 2021-04-08 00:35:10.168 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.Core.Twin.StoringTwinManager] - Getting twin for MTC01_DATAREADER
<7> 2021-04-08 00:35:10.168 +00:00 [DBG] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Creating cloud connection for client MTC01_DATAREADER using EdgeHub credentials
<6> 2021-04-08 00:35:10.168 +00:00 [INF] [Microsoft.Azure.Devices.Edge.Hub.CloudProxy.CloudConnection] - Attempting to connect to IoT Hub for client MTC01_DATAREADER via AMQP...
```
</details>

## Additional Information
Something of note is that if the leaf device is already connected to the edge device when the edge device loses internet, there is no degradation of service. However, if after that point the leaf device restarts, for example, and thereby tries to reconnect, this issue occurs (this is essentially what I described in repro steps above).

Additionally, I can unreliably get the leaf device's messages flowing while the IoT Edge Gateway remains offline. However, if I knew how to cause this behavior reliably, I may not be creating this issue.

Thanks for your help.

**Comment by vadim-kovalyov** [comment]
@jackt-moran, thanks for reporting this, I'll try to reproduce it myself and report back. It seems that I have enough info for now to try it myself. The only question I have is what SDK and version do you use on leaf device?

**Comment by jackt-moran** [comment]
@vadim-kovalyov, I started working on gathering the leaf device's logs (we're using the Python SDK 2.5.1), and it looks like I may have resolved the worst of the issue through a change I made while troubleshooting. At this point, edgeHub still blocks the processing of leaf device messages until it gives up trying to establish a cloud connection, but once it gives up message processing resumes as normal.

Here's the change that I made the I believe was causing the issue (please let me know if this sounds plausible):
- My old leaf device code first established a twin patch handler, then issued a GET for it's initial twin.
- I saw in another issue in this repository the operation ordering of the above bullet could be problematic, so as a shot in the dark I switched the order of those two operations. Now my leaf device issues an initial GET, and only after that operation completes do I define the twin patch handler. Could this change fix the root cause?

Assuming everything I'm saying is correct and the behavior I'm seeing is expected, I would argue that the blocking behavior is still undesirable. It is of course not as critical as the initial issue I was experiencing, but offline operation is expected in our use case which is part of the reason why we're using IoT Edge. Given that the connection between my leaf device and gateway is via LAN, I'd want minimal or no downtime regardless of if the gateway has internet or not.

**Comment by vadim-kovalyov** [comment]
@jackt-moran, Hm... your comment above leads me to believe that the issue might be in the Python SDK. That make sense, b/c edgeHub should not block incoming packet stream on any operations. I'm not an expert in Python, but AFAIK it by default does not have concurrent code execution, so it is possible that some order of operation creates blocking. Is it possible for you to get a simple PoC module code to repro the issue? And with that we can probably ask SDK team for help.

**Comment by vadim-kovalyov** [comment]
Hey @jackt-moran Is it possible for you to share a simple PoC module code to repro the issue?

**Comment by jackt-moran** [comment]
Hey @vadim-kovalyov sorry for the delay - yes I can provide PoC, device twin to repro the issue. If you'd prefer that I verify the code repros before sharing it, I will need some time to verify that it repros (about a week from now - just a lot going on with my project team right now). If you're willing/able to attempt to repro yourself, I could share some PoC code whenever. Let me know what you'd prefer, thanks.

**Comment by vadim-kovalyov** [comment]
@jackt-moran please provide whatever you have so I can give it a try. If I'm not able to repro, I'll wait for your working PoC. If I'm able to repro that, we'll forward that to SDK team.

**Comment by jackt-moran** [comment]
@vadim-kovalyov sounds good to me. I included below a script that _should_ repro, if it was indeed the order of API calls that exposed the issue. You'll see that the device twin patch handler is explicitly defined before any call to get_twin() is made, which is what I was describing above. Let me know if you have any issues with the PoC (azure-iot-device==2.5.1, Python 3.7, OS: Ubuntu 18.04):

```python
import asyncio
import datetime
import os
import random
import traceback

from azure.iot.device.aio import IoTHubDeviceClient
from azure.iot.device.exceptions import (
    CredentialError,
    ConnectionDroppedError,
    ConnectionFailedError,
    ClientError,
)


async def connect_client(module_client: IoTHubDeviceClient):
    """
    Connects a Module/Device Client to the IoT Edge runtime's Edge Hub.
    This coroutine is interruptible, but will otherwise attempt
    to connect forever until successful.

    Returns:
        None.
    """
    connected = False
    while not connected:
        try:
            await module_client.connect()
            connected = True
        except CredentialError:
            traceback.print_exc()
            raise
        except (ConnectionFailedError, ConnectionDroppedError):
            traceback.print_exc()
            print("Attempting to connect again")
        except ClientError:
            traceback.print_exc()
            raise
        await asyncio.sleep(1)


async def dummy_produce(device_client: IoTHubDeviceClient, delay: int, coro_id: int):
    while True:
        await device_client.send_message(
            {
                "data": random.random(),
                "coro_id": coro_id,
                "timestamp": datetime.datetime.now().isoformat(),
            }
        )
        await asyncio.sleep(delay)


async def run():
    with open(os.environ["EDGEHUB_PUBLIC_ROOT_CA_PATH"]) as f:
        edgehub_ca_cert = f.read()
    edgehub_connection_string = os.environ["EDGEHUB_CONNECTION_STRING"]
    device_client = IoTHubDeviceClient.create_from_connection_string(
        edgehub_connection_string,
        server_verification_cert=edgehub_ca_cert,
    )
    await connect_client(device_client)

    async def twin_patch_handler(patch: dict):
        print("Received twin update: {}".format(patch))

    # set the twin patch handler on the client
    device_client.on_twin_desired_properties_patch_received = twin_patch_handler

    initial_twin = await device_client.get_twin()
    print("Received initial twin: {}".format(initial_twin))

    num_producers = 16
    delay_seconds = 1
    producers = [
        dummy_produce(device_client, delay_seconds, i) for i in range(num_producers)
    ]
    # Event volume should consistently be 960/minute or 16/second.
    await asyncio.gather(*producers)


# Note: this script requires the env vars EDGEHUB_PUBLIC_ROOT_CA_PATH
# and EDGEHUB_CONNECTION_STRING to be defined.
if __name__ == "__main__":
    # Requires Python 3.7+
    asyncio.run(run())

```

**Comment by vadim-kovalyov** [comment]
@jackt-moran, thanks for the code. I have played with and here is what I found. Indeed, there is a possibility to encounter the situation that you have described. I was able to repro with Python and dotnet SDKs.

TL;DR;

There is an internal state machine in EdgeHub for upstream connections. When EdgeHub encounters network problems with the upstream, it has ~60sec before AMQP connection times out and closes. During that time the connection is in `Unreachable` state. If at this time a client tries to perform "get twin" operation, it will trigger "upstream connection check" which, in turn, has also 60s default timeout. But if network is down for a while (>60s). The EdgeHub goes into `Disconnected` state. And when in `Disconnected` state, every iothub-specific operation (like get twin) is bound by `CloudOperationTimeoutSecs` setting (default 20s).

So, if device happened to connect right after EdgeHub went offline and before AMQP connection timed out - then it will get stuck with 60s "upstream connection check" timeout. If device connects a bit later - then the timeout is bound by `CloudOperationTimeoutSecs` setting. `CloudOperationTimeoutSecs` can be changed via env var.

I'll bring this issue to the team's attention to discuss the fix. Meanwhile, my recommendation as a workaround: Since `get_twin()` is a "blocking call" anyways, I would suggest to run it in a separate background task. And let me ask you this - do you even need to call `get_twin()`? Isn't callback enough for your purposes? What is your scenario for `get_twin()`?

**Comment by jackt-moran** [comment]
@vadim-kovalyov thank you for the detailed explanation.

I can appreciate that changing the behavior you've explained could be difficult, so in the meantime I will experiment with skipping the call to `get_twin()`. Depending on the behavior of the callback, making a call to `get_twin()` may be a requirement. 

Our device code internally manages persisting it's twin, largely motivated by part of our config (large files) being stored in a data lake and referenced in our twins. In general, due to this persistence, a `get_twin()` call is unnecessary. However, for a device that's running for the first time it will not have a local version of it's twin. In this case, I assumed a `get_twin()` call would be required to retrieve the twin initially.

I wanted to avoid the scenario where a device is turned on for the first time, receives no twin updates for an arbitrary period of time (and therefore no callback is triggered), and as a result has no twin and can do nothing for this period of time - if we call `get_twin()` explicitly at start-up this scenario should be impossible. However, if for example upon registering a callback that callback is guaranteed to be triggered, then the "failure" scenario I described would be avoided as well. Perhaps my assumptions were incorrect about the behavior of the callback?

**Comment by vadim-kovalyov** [comment]
@jackt-moran, `on_twin_desired_properties_patch_received` callback is non-blocking (if you don't have any blocking code inside the callback handler, of course), and it will be called only when desired properties have changed in iothub. So, if I understand your scenario right, here is the logic:
```rust
if (persisted_state) { /// this case will not block, since no blocking `get_twin()` call
    register_on_twin_change(handler_func)
    load_state_and_run(persisted_state)
} else { /// this case can block if `get_twin()` takes a lot of time. 
    register_on_twin_change(handler_func)
    state = get_twin() /// if device connects for the first time and EdgeHub is offline, this can block for up to 60s and will be empty. I don't see a way to recover out of that, except to bring EdgeHub back online.
    load_state_and_run(state)
}

//cc @varunpuranik to keep me honest on `on_twin_desired_properties_patch_received` behavior.

**Comment by jackt-moran** [comment]
@vadim-kovalyov thanks for your explanation here. What you shared is essentially what I came up with as well before hearing back from you. This seems to work well on our production deployments so overall it looks like a good workaround/solution.

I would still argue this is something of a workaround (of course that can be up for debate) as we have some smoke testing that validates (desired props) == (reported props) for our devices. If a device is disconnected or off when a property update is sent, it seems possible that it would never receive the update (I don't know this positively; I assume this behavior depends on how edgeHub handles twin patches that occurred while edgeHub is offline then goes from disconnected to connected). With your code as a guide, it seems possible that a device with `persisted_state` defined, that is offline when a twin patch is created, that later comes online, would never receive the twin patch (because the handler wasn't available when the patch was created) and instead would just load the `persisted_state` rather than calling `get_twin()`. In general this isn't a big problem for us, but it does seem to present us with a tradeoff of application uptime vs. ease of configurability/configuration testing. 

Do I have the gist of what I describe correct, and if so do you see what I mean when I say this is a workaround rather than a solution? Also keep in mind, seeing as what you shared above does solve the immediate problem, I wouldn't call this a major issue for my use case.

**Comment by vadim-kovalyov** [comment]
@jackt-moran, you're totally right that in my example above it is possible to miss the update if device goes offline for some time. I should have thought about that, but I missed that point.. it is actually described in [our twin docs](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-devguide-device-twins#device-reconnection-flow). So, to make the solution complete you would need to still query twin when device comes back online even if the persistent state is present. But to avoid blocking if you already have some state to start with, you should run `get_twin()` in a background task. It will require some synchronization, but in my opinion this seems like most reliable solution that minimizes startup delays.

Just one more FYI: if device is connected to EdgeHub, and EdgeHub goes offline for some time, and then back online, it (EdgeHub) will re-query twins on re-connect and will send delta back to the device, so for the device it should transparent and it should not lose any updates.

**Comment by vadim-kovalyov** [comment]
I'm closing this for now. Please mention me if you have any further question on this issue.
